doc/developers/develop.rst

```diff
@@ -81,33 +92,35 @@ classifier or a regressor. All estimators implement the fit method::
 
     estimator.fit(X, y)
 
-All built-in estimators also have a ``set_params`` method, which sets
-data-independent parameters (overriding previous parameter values passed
-to ``__init__``).
-
-All estimators in the main scikit-learn codebase should inherit from
-``sklearn.base.BaseEstimator``.
+Out of all the methods that an estimator implements, ``fit`` is usually the one you
+want to implement yourself. Other methods such as ``set_params``, ``get_params``, etc.
+are implemented in :class:`~sklearn.base.BaseEstimator`, which you should inherit from.
+You might need to inherit from more mixins, which we will explain later.
 
 Instantiation
 ^^^^^^^^^^^^^
 
-This concerns the creation of an object. The object's ``__init__`` method
-might accept constants as arguments that determine the estimator's behavior
-(like the C constant in SVMs). It should not, however, take the actual training
-data as an argument, as this is left to the ``fit()`` method::
+This concerns the creation of an object. The object's ``__init__`` method might accept
+constants as arguments that determine the estimator's behavior (like the ``alpha``
+constant in :class:`~sklearn.linear_model.SGDClassifier`). It should not, however, take
+the actual training data as an argument, as this is left to the ``fit()`` method::
 
-    clf2 = SVC(C=2.3)
-    clf3 = SVC([[1, 2], [2, 3]], [-1, 1]) # WRONG!
+    clf2 = SGDClassifier(alpha=2.3)
+    clf3 = SGDClassifier([[1, 2], [2, 3]], [-1, 1]) # WRONG!
 
 
-The arguments accepted by ``__init__`` should all be keyword arguments
-with a default value. In other words, a user should be able to instantiate
-an estimator without passing any arguments to it. The arguments should all
-correspond to hyperparameters describing the model or the optimisation
-problem the estimator tries to solve. These initial arguments (or parameters)
-are always remembered by the estimator.
-Also note that they should not be documented under the "Attributes" section,
-but rather under the "Parameters" section for that estimator.
+Ideally, the arguments accepted by ``__init__`` should all be keyword arguments with a
+default value. In other words, a user should be able to instantiate an estimator without
+passing any arguments to it. In some cases, where there are no sane defaults for an
+argument, they can be left without a default value. In scikit-learn itself, we have
+very few places, only in some meta-estimators, where the sub-estimator(s) argument is
+a required argument.
+
+The arguments should all correspond to hyperparameters describing the model or the
```

Comment:
Not all arguments are hyparmeters or part of the optimization problem. For example `CountVectorizer.input` is not in a search space for tuning a model.

Reply:
This was text from before, but I agree, changed now.